File: server/src/services/categorizer.ts
import prisma from "../db";

// Keyword-based auto-categorization rules for HDFC transactions
const KEYWORD_RULES: { pattern: RegExp; categoryName: string }[] = [
  // Investments
  { pattern: /GROWW|STOCKSIP|DEBITFORSTOCKSS/i, categoryName: "Investment - SIP" },
  { pattern: /MUTUALFUNDS|INDIANCLEARING/i, categoryName: "Investment - Mutual Fund" },
  { pattern: /NETBANKINGSI-PPF|PPF/i, categoryName: "Investment - PPF" },
  { pattern: /RDINSTALLMENT/i, categoryName: "Investment - RD" },

  // Transfers
  { pattern: /CC\d+.*AUTOPAY|AUTOPAY.*THANK\s*YOU/i, categoryName: "Credit Card Payment" },

  // Income
  { pattern: /INTERESTPAID|INTEREST PAID/i, categoryName: "Interest" },
  { pattern: /TRAVELSTACK/i, categoryName: "Salary" },
  { pattern: /BAINCAP|BAIN COMPANY/i, categoryName: "Salary" },
  { pattern: /URBANCLAP|URBAN COMPANY/i, categoryName: "Salary" },
  { pattern: /ACH\s+C-.*SAL/i, categoryName: "Salary" },
  { pattern: /CDSL|NEFTCR.*CENTRALDEPOSITOR/i, categoryName: "Dividends" },
  { pattern: /ACHC-RAILVIKAS/i, categoryName: "Dividends" },
  { pattern: /ACHD-INDIANCLEARING/i, categoryName: "Investment - Mutual Fund" },

  // Expenses
  { pattern: /NWD-|ATW-/i, categoryName: "ATM Withdrawal" },
  { pattern: /POS\s+\d/i, categoryName: "Misc Expense" },
  { pattern: /RENT|TPT-.*RENT/i, categoryName: "Rent" },
  { pattern: /ZOMATO|SWIGGY|PYU\*ZOMATO/i, categoryName: "Food Delivery" },
  { pattern: /BLINKIT|ZEPTO|BIGBASKET|INSTAMART/i, categoryName: "Groceries" },
  { pattern: /FLIPKART|AMAZON(?!.*PAY.*PRIV.*EMI)/i, categoryName: "Shopping" },
  { pattern: /UBER|OLA|INDIGO.*AIRLINE|RAPIDO/i, categoryName: "Transport" },
  { pattern: /BIGTREE|BOOKMYSHOW/i, categoryName: "Entertainment" },
  { pattern: /CLAUDE\.AI|GOOGLE\s*PLAY|F1\.COM|NETFLIX|SPOTIFY|HOTSTAR/i, categoryName: "Subscriptions" },
  { pattern: /MILLENNIUM.*HEALT|PHARMACY|MEDIC|THEMILLENNIUM/i, categoryName: "Health" },
  { pattern: /KEVENTER|BHAGWATI|RANI\s*KUMARI|JEETENDRA/i, categoryName: "Groceries" },
  { pattern: /NEFTDR-.*SHANTANU|NEFTDR-.*HDFCH/i, categoryName: "Credit Card Payment" },
  { pattern: /UPI-SURABHI/i, categoryName: "Misc Expense" },
  { pattern: /UPI-AMLAN/i, categoryName: "Misc Expense" },
  { pattern: /UPI-MANSI/i, categoryName: "Misc Expense" },
  { pattern: /UPI-RANI/i, categoryName: "Groceries" },

  // CC specific
  { pattern: /IGST-|CGST-|SGST-|FCY\s*MARKUP\s*FEE|CONSOLIDATED\s*FCY/i, categoryName: "Fees/Charges" },
  { pattern: /EMI\s+AMAZON\s*PAY|EMI\s+ANJEER|EMI\s+BIGTREE|EMI\s+WWW\.F1/i, categoryName: "Entertainment" },
];

// Cache category name → id mapping
let categoryCache: Record<string, string> | null = null;

async function getCategoryMap(): Promise<Record<string, string>> {
  if (categoryCache) return categoryCache;
  const categories = await prisma.category.findMany();
  categoryCache = {};
  for (const c of categories) {
    categoryCache[c.name] = c.id;
  }
  return categoryCache;
}

export function invalidateCategoryCache() {
  categoryCache = null;
}

export async function autoCategorize(description: string, counterparty?: string): Promise<{ categoryId: string | null; categoryName: string | null }> {
  const catMap = await getCategoryMap();
  const textToMatch = counterparty ? `${description} ${counterparty}` : description;

  for (const rule of KEYWORD_RULES) {
    if (rule.pattern.test(textToMatch)) {
      const categoryId = catMap[rule.categoryName];
      if (categoryId) {
        return { categoryId, categoryName: rule.categoryName };
      }
    }
  }

  return { categoryId: null, categoryName: null };
}

File: server-py/app/services/categorizer.py
"""Keyword-based auto-categorization — faithful port of server/src/services/categorizer.ts.

Behavior preserved exactly (regex + first-match-wins). The category-name->id map
is cached and can be invalidated when categories change.
"""
import re
from sqlmodel import Session, select
from ..models import Category

# (pattern, category_name) — ported 1:1 from categorizer.ts KEYWORD_RULES
_RAW_RULES: list[tuple[str, str]] = [
    # Investments
    (r"GROWW|STOCKSIP|DEBITFORSTOCKSS", "Investment - SIP"),
    (r"MUTUALFUNDS|INDIANCLEARING", "Investment - Mutual Fund"),
    (r"NETBANKINGSI-PPF|PPF", "Investment - PPF"),
    (r"RDINSTALLMENT", "Investment - RD"),
    # Transfers
    (r"CC\d+.*AUTOPAY|AUTOPAY.*THANK\s*YOU", "Credit Card Payment"),
    # Income
    (r"INTERESTPAID|INTEREST PAID", "Interest"),
    (r"TRAVELSTACK", "Salary"),
    (r"BAINCAP|BAIN COMPANY", "Salary"),
    (r"URBANCLAP|URBAN COMPANY", "Salary"),
    (r"ACH\s+C-.*SAL", "Salary"),
    (r"CDSL|NEFTCR.*CENTRALDEPOSITOR", "Dividends"),
    (r"ACHC-RAILVIKAS", "Dividends"),
    (r"ACHD-INDIANCLEARING", "Investment - Mutual Fund"),
    # Expenses
    (r"NWD-|ATW-", "ATM Withdrawal"),
    (r"POS\s+\d", "Misc Expense"),
    (r"RENT|TPT-.*RENT", "Rent"),
    (r"ZOMATO|SWIGGY|PYU\*ZOMATO", "Food Delivery"),
    (r"BLINKIT|ZEPTO|BIGBASKET|INSTAMART", "Groceries"),
    (r"FLIPKART|AMAZON(?!.*PAY.*PRIV.*EMI)", "Shopping"),
    (r"UBER|OLA|INDIGO.*AIRLINE|RAPIDO", "Transport"),
    (r"BIGTREE|BOOKMYSHOW", "Entertainment"),
    (r"CLAUDE\.AI|GOOGLE\s*PLAY|F1\.COM|NETFLIX|SPOTIFY|HOTSTAR", "Subscriptions"),
    (r"MILLENNIUM.*HEALT|PHARMACY|MEDIC|THEMILLENNIUM", "Health"),
    (r"KEVENTER|BHAGWATI|RANI\s*KUMARI|JEETENDRA", "Groceries"),
    (r"NEFTDR-.*SHANTANU|NEFTDR-.*HDFCH", "Credit Card Payment"),
    (r"UPI-SURABHI", "Misc Expense"),
    (r"UPI-AMLAN", "Misc Expense"),
    (r"UPI-MANSI", "Misc Expense"),
    (r"UPI-RANI", "Groceries"),
    # CC specific
    (r"IGST-|CGST-|SGST-|FCY\s*MARKUP\s*FEE|CONSOLIDATED\s*FCY", "Fees/Charges"),
    (r"EMI\s+AMAZON\s*PAY|EMI\s+ANJEER|EMI\s+BIGTREE|EMI\s+WWW\.F1", "Entertainment"),
]

KEYWORD_RULES = [(re.compile(p, re.IGNORECASE), name) for p, name in _RAW_RULES]

_category_cache: dict[str, str] | None = None


def get_category_map(session: Session) -> dict[str, str]:
    global _category_cache
    if _category_cache is not None:
        return _category_cache
    cats = session.exec(select(Category)).all()
    _category_cache = {c.name: c.id for c in cats}
    return _category_cache


def invalidate_category_cache() -> None:
    global _category_cache
    _category_cache = None


def auto_categorize(session: Session, description: str,
                    counterparty: str | None = None) -> str | None:
    """Return a categoryId or None. Mirrors categorizer.ts autoCategorize."""
    cat_map = get_category_map(session)
    text = f"{description} {counterparty}" if counterparty else (description or "")
    for pattern, name in KEYWORD_RULES:
        if pattern.search(text):
            cid = cat_map.get(name)
            if cid:
                return cid
    return None
